### PR TITLE
V22F-799 Options do not save when navigate to sub-menu

### PR DESCRIPTION
**Pull request by LucasDissie**

### CODE REVIEW

**Review by Sindarin27 — DISMISSED** [review]
When going into the Force Feedback menu in the Controls options, changed controls will still be reloaded.

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp [hunk 1]
```diff
@@ -661,6 +649,19 @@ void* ResearcherMenuInit(void* p_nextMenu)
...
     GfuiEditboxSetString(s_scrHandle, m_userIdControl, m_userId);
 
+    // Retrieves the saved user xml file, if it doesn't exist the settings are already initialized in ResearcherMenuInit
+    char buf[512];
+    sprintf(buf, "%s%s", GfLocalDir(), RESEARCH_FILEPATH);
+    if (GfFileExists(buf))
+    {
+        void* param = GfParmReadFile(buf, GFPARM_RMODE_STD);
+        // Initialize settings with the retrieved xml file
+        LoadConfigSettings(param);
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp:659]: The force feedback option can also be changed in the force feedback menu. Changing it there doesn't update it in the researchermenu.

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseConnectionManager.cpp [hunk 2]
```diff
@@ -198,35 +195,44 @@ void AsyncCheckConnection(void* p_scrHandle, int p_dbStatusControl, tDatabaseSet
...
-    GfuiEditboxSetString(p_scrHandle, p_usernameControl, s_dbSettings.Username);
+    strcpy_s(s_tempDbSettings.Username, SETTINGS_NAME_LENGTH, GfuiEditboxGetString(p_scrHandle, p_usernameControl));
+    GfuiEditboxSetString(p_scrHandle, p_usernameControl, s_tempDbSettings.Username);
 }
 
 /// @brief Handle input in the Password textbox
 /// @param p_scrHandle The screen handle which to operate the functions on
 /// @param p_passwordControl the corresponding ui element control integers
-void SetPassword(void* p_scrHandle, int p_passwordControl)
+void SetPassword(void* p_scrHandle, int p_passwordControl, char* p_password)
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseConnectionManager.cpp:232]: missing param p_password in documentation

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseConnectionManager.cpp [hunk 3]
```diff
@@ -236,13 +242,24 @@ void SetPassword(void* p_scrHandle, int p_passwordControl)
...
+void FillInPassword(void* p_scrHandle, int p_passwordControl)
+{
+    SetPassword(p_scrHandle, p_passwordControl, s_dbSettings.Password);
+}
+
+void ChangePassword(void* p_scrHandle, int p_passwordControl)
+{
+    strcpy_s(s_tempDbSettings.Password, SETTINGS_NAME_LENGTH, GfuiEditboxGetString(p_scrHandle, p_passwordControl));
+    SetPassword(p_scrHandle, p_passwordControl, s_tempDbSettings.Password);
+}
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseConnectionManager.cpp:254]: missing documentation

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseConnectionManager.cpp [hunk 4]
```diff
@@ -312,7 +329,7 @@ void SelectCert(void* p_scrHandle, int p_buttonControl, int p_labelControl, cons
     }
 
     // Validate input w.r.t. black boxes
-    filesystem::path path = buf;
+    std::experimental::filesystem::path path = buf;
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseConnectionManager.cpp:332]: std::experimental:: is unneeded
```suggestion
    filesystem::path path = buf;
```

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseSettingsMenu.cpp [hunk 5]
```diff
@@ -94,7 +95,7 @@ static void SetUsernameCallback(void*)
 
 static void SetPasswordCallback(void*)
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseSettingsMenu.cpp:96]: documentation

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseSettingsMenu.cpp [hunk 6]
```diff
@@ -94,7 +95,7 @@ static void SetUsernameCallback(void*)
...
 {
-    SetPassword(s_scrHandle, m_passwordControl);
+    ChangePassword(s_scrHandle, m_passwordControl);
 }
 
 static void SetAddressCallback(void*)
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseSettingsMenu.cpp:101]: documentation

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DataSelectionMenu.cpp [hunk 7]
```diff
@@ -106,23 +106,7 @@ static void LoadDefaultSettings()
 /// @brief Loads the user menu settings from the local config file
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DataSelectionMenu.cpp:106]: @brief is lying

**Review by shoekzema — CHANGES_REQUESTED** [review]

**Review by suppergerrie2 — CHANGES_REQUESTED** [review]
Add to Code Changes

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseConnectionManager.cpp [hunk 8]
```diff
@@ -198,35 +195,47 @@ void AsyncCheckConnection(void* p_scrHandle, int p_dbStatusControl, tDatabaseSet
...
+/// @param  p_dbStatusControl The status control handle to write letters to the screen
+/// @param  p_isConnecting    Checks if a connection is already being made
+void CheckCurrentConnection(void* p_scrHandle, int p_dbStatusControl, bool* p_isConnecting)
+{
+    if (*p_isConnecting) return;
+    *p_isConnecting = true;
+    std::thread t(AsyncCheckConnection, p_scrHandle, p_dbStatusControl, s_tempDbSettings, p_isConnecting);
+    t.detach();
+}
+
```
[the remark this answers is not in the document]
Comment (LucasDissie) [reply speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseConnectionManager.cpp:221]: The reason i did split them is because these functions are called outside the class where i have no access to s_tempDbSettings and s_DbSettings. I will make one function with an extra argument and just call those in the other functions.

**Review by larsdekwant — CHANGES_REQUESTED** [review]
- The check connection function is now extremely slow, this should be a responsive label.
-  Move the error label down so that its more centered in the gap between choose environment and the back button (label is shown by pressing next without generating an ID)

File: speed-dreams/source-2.2.3/data/data/menu/controlconfigmenu.xml [hunk 9]
```diff
@@ -453,29 +453,6 @@
         <attstr name="pushed color" val="0xFFCC00"/>
     </section>
 
-    <!-- SIMULATED DRIVING ASSISTANCE-->
-    <section name="ForceFeedbackButton">
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/data/data/menu/controlconfigmenu.xml:457]: remove from CHANGES.txt

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/confscreens/controlconfig.cpp [hunk 10]
```diff
@@ -41,15 +41,6 @@
 
 #include <deque>
 
-// SIMULATED DRIVING ASSISTANCE CHANGE: added include for force feedback config and defines
-#if SDL_FORCEFEEDBACK
-#include <forcefeedbackconfig.h>
-#define HUMAN_DRIVER_INDEX        1
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/confscreens/controlconfig.cpp:47]: remove from CHANGES.txt (the two deleted parts below too)

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp [hunk 11]
```diff
@@ -70,6 +71,14 @@
 
 #define TRACK_LOADER_MODULE_NAME "trackv1"
 
+// SIMULATED DRIVING ASSISTANCE CHANGE: added include for force feedback config and defines
+#if SDL_FORCEFEEDBACK
+#include <forcefeedbackconfig.h>
+#define HUMAN_DRIVER_INDEX        1
+#define PRIMO_CAR_NAME            "primo"
+#define PRM_FORCE_FEEDBACK_BUTTON "ForceFeedbackButton"
+#endif
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp:80]: change comment in our own file?

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/ConfigTests.cpp [hunk 12]
```diff
@@ -71,14 +71,13 @@ END_TEST_COMBINATORIAL3(IndicatorTest, booleans, 2, booleans, 2, booleans, 2)
 void PControlTest(bool p_steer, bool p_gas, bool p_brake, bool p_intervention, bool p_force)
 {
     SDAConfig config;
-    tParticipantControl arr = {p_steer, p_gas, p_brake, p_intervention, p_force};
+    tParticipantControl arr = {p_steer, p_gas, p_brake, p_intervention};
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/ConfigTests.cpp:74]: now p_force is unused, so remove it (from function, documentation and usages of function)

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 13]
```diff
@@ -197,24 +197,23 @@ END_TEST_COMBINATORIAL3(IndicatorTestMediator, booleans, 2, booleans, 2, boolean
 /// @param p_gas          Control gas option
 /// @param p_steer        Control steering option
 /// @param p_force        Force feedback option
-void PControlTestMediator(bool p_steer, bool p_gas, bool p_brake, bool p_intervention, bool p_force)
+void PControlTestMediator(bool p_steer, bool p_gas, bool p_brake, bool p_intervention)
 {
     SDAConfigMediator::ClearInstance();
     ASSERT_TRUE(SetupSingletonsFolder());
-    tParticipantControl arr = {p_steer, p_gas, p_brake, p_intervention, p_force};
+    tParticipantControl arr = {p_steer, p_gas, p_brake, p_intervention};
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:204]: now p_force is unused, so remove it (from function, documentation and usages of function)

**Review by shoekzema — CHANGES_REQUESTED** [review]

**Review by larsdekwant — CHANGES_REQUESTED** [review]
Label is stuck on `Connecting...` for a long time. 
The port is always 0 if you don't explicitly 'edit' the port box.
These two issues might be related? (as in that connecting might take a long time to a non-existent port)

**Review by shoekzema — CHANGES_REQUESTED** [review]

**Review by larsdekwant — CHANGES_REQUESTED** [review]
The long connecting bug is fixed now that the port is no longer always 0, however now it is not possible to edit the port on the menu. Whenever I stop editing, the value is reverted to what was in the editbox before.

**Review by shoekzema — APPROVED** [review]

**Review by suppergerrie2 — APPROVED** [review]
